mpas_analysis/shared/climatology/climatology.py

```diff
@@ -386,94 +385,104 @@ def get_observation_climatology_file_names(config, fieldName, monthNames,
+def compute_climatology(ds, monthValues, calendar=None):
     """
-    Given a monthly climatology, compute a seasonal climatology weighted by
-    the number of days in each month (on the no-leap-year calendar).
+    Compute a monthly, seasonal or annual climatology data set from a data
+    set.  The mean is weighted but the number of days in each month of
+    the data set, ignoring values masked out with NaNs.  If the month
+    coordinate is not present, a data array ``month`` will be added based
+    on ``Time`` and the provided calendar.
 
     Parameters
     ----------
-    monthlyClimatology : instance of xarray.DataSet
-        A data set containing a monthly climatology
+    ds : ``xarray.Dataset`` or ``xarray.DataArray`` object
+        A data set with a ``Time`` coordinate expressed as days since
+        0001-01-01 or ``month`` coordinate
 
     monthValues : int or array-like of ints
         A single month or an array of months to be averaged together
-        before interpolation.
+
+    calendar: ``{'gregorian', 'gregorian_noleap'}``, optional
+        The name of one of the calendars supported by MPAS cores, used to
+        determine ``month`` from ``Time`` coordinate, so must be supplied if
+        ``ds`` does not already have a ``month`` coordinate or data array
 
     Returns
     -------
-    seasonalClimatology : instance of xarray.DataSet
-        A data set containing the seasonal climatology
+    climatology : object of same type as ``ds``
+        A data set without the ``'Time'`` coordinate containing the mean
+        of ds over all months in monthValues, weighted by the number of days
+        in each month.
 
     Authors
     -------
     Xylar Asay-Davis
 
     Last Modified
     -------------
-    02/27/2017
+    03/30/2017
     """
 
-    # select only the desired field from the obs. data set
-    monthlyClimatology = mpas_xarray.subset_variables(
-        monthlyClimatology, variableList=[variableName])
+    if ('month' not in ds.coords or 'daysInMonth' not in ds.coords):
+        ds = add_months_and_days_in_month(ds, calendar)
 
-    # Make a DataArray with the number of days in each month
-    daysInMonth = xr.DataArray(numpy.array(constants.daysInMonth, float),
-                               coords=[monthlyClimatology.month],
-                               name='daysInMonth')
+    mask = xr.zeros_like(ds.month, bool)
 
-    daysInMonth = daysInMonth.sel(month=monthValues)
-    monthlyClimatology = monthlyClimatology.sel(month=monthValues)
+    for month in monthValues:
+        mask = xr.ufuncs.logical_or(mask, ds.month == month)
 
-    varArray = monthlyClimatology[variableName]
-    mask = xr.DataArray(~numpy.isnan(varArray.values),
-                        coords=varArray.coords,
-                        name='mask')
-    seasonalClimatology = (monthlyClimatology * daysInMonth).sum(
-        dim='month', keep_attrs=True)
+    climatologyMonths = ds.where(mask, drop=True)
```

Comment:
Nice!  This should help below when `_compute_masked_mean` is used for the computation. :+1: